src/ng/directive/ngOptions.js

```diff
@@ -686,53 +650,32 @@ var ngOptionsDirective = ['$compile', '$parse', function($compile, $parse) {
 
             if (!group) {
 
-              // We have not already created this group
-              groupElement = addOrReuseElement(selectElement[0],
-                                               currentElement,
-                                               'optgroup',
-                                               optGroupTemplate);
-              // Move to the next element
-              currentElement = groupElement.nextSibling;
+              groupElement = optGroupTemplate.cloneNode(false);
+              listFragment.appendChild(groupElement);
 
               // Update the label on the group element
               groupElement.label = option.group;
 
               // Store it for use later
               group = groupMap[option.group] = {
-                groupElement: groupElement,
-                currentOptionElement: groupElement.firstChild
+                groupElement: groupElement
               };
 
             }
 
-            // So now we have a group for this option we add the option to the group
-            optionElement = addOrReuseElement(group.groupElement,
-                                              group.currentOptionElement,
-                                              'option',
-                                              optionTemplate);
+            optionElement = optionTemplate.cloneNode(false);
+            group.groupElement.appendChild(optionElement);
             updateOptionElement(option, optionElement);
-            // Move to the next element
-            group.currentOptionElement = optionElement.nextSibling;
 
           } else {
 
-            // This option is not in a group
-            optionElement = addOrReuseElement(selectElement[0],
-                                              currentElement,
-                                              'option',
-                                              optionTemplate);
+            optionElement = optionTemplate.cloneNode(false);
```

Comment:
Nitpick: this path is almost identical with above (only the parentElement changes). We could factor it out of the `if-else` block.